File: VerbalExpressions.java
import java.util.regex.Pattern;

class VerbalExpressions {
    private String prefixes, source, suffixes, pattern = "";
    int modifiers = Pattern.MULTILINE;
    Pattern p;

    public VerbalExpressions() {

    }

    private String sanitize(String value) {
        if(value != null) return value;
        return Pattern.quote(value);
    }

    public VerbalExpressions add(String value) {
        this.source = this.source != null ? this.source + value : value;
        if (this.source != null) {
            this.p = Pattern.compile(this.prefixes + this.source + this.suffixes, this.modifiers);
            this.pattern = this.p.pattern();
        }
        return this;
    }

    public VerbalExpressions startOfLine(boolean enable) {
        this.prefixes = enable ? "^" : "";
        this.add( "" );
        return this;
    }

    public VerbalExpressions startOfLine() {
        return startOfLine(true);
    }

    public VerbalExpressions endOfLine(boolean enable) {
        this.suffixes = enable ? "$" : "";
        this.add( "" );
        return this;
    }

    public VerbalExpressions endOfLine() {
        return endOfLine(true);
    }

    public VerbalExpressions then(String value) {
        value = sanitize(value);
        this.add( "(" + value + ")" );
        return this;
    }

    public VerbalExpressions find(String value) {
        this.then(value);
        return this;
    }

    public VerbalExpressions maybe(String value) {
        value = sanitize(value);
        this.add( "(" + value + ")?" );
        return this;
    }

    public VerbalExpressions anything() {
        this.add( "(.*)" );
        return this;
    }

    public VerbalExpressions anythingBut(String value) {
        value = sanitize(value);
        this.add( "([^" + value + "]*)" );
        return this;
    }

    public VerbalExpressions replace(String source, String value) {
        this.add( "" );
        this.source.replaceAll(pattern,value);
        return this;
    }

    public VerbalExpressions lineBreak() {
        this.add( "(\\n|(\\r\\n))" );
        return this;
    }

    public VerbalExpressions br() {
        this.lineBreak();
        return this;
    }

    public VerbalExpressions tab() {
        this.add( "\\t" );
        return this;
    }

    public VerbalExpressions word() {
        this.add( "\\w+" );
        return this;
    }

    public VerbalExpressions anyOf(String value) {
        value = sanitize(value);
        this.add( "[" + value + "]" );
        return this;
    }

    public VerbalExpressions any(String value) {
        this.anyOf(value);
        return this;
    }

    public VerbalExpressions range(Object[] args) {
        String value = "[";
        for(int _from = 0; _from < args.length; _from += 2) {
            int _to = _from+1;
            if (args.length <= _to) break;
            int from = Integer.getInteger(sanitize((String)args[_from]));
            int to = Integer.getInteger(sanitize((String)args[_to]));

            value += from + "-" + to;
        }

        value += "]";

        this.add(value);
        return this;
    }

    public VerbalExpressions addModifier(char modifier) {
        switch (modifier) {
            case 'd':
                modifiers |= Pattern.UNIX_LINES;
                break;
            case 'i':
                modifiers |= Pattern.CASE_INSENSITIVE;
                break;
            case 'x':
                modifiers |= Pattern.COMMENTS;
                break;
            case 'm':
                modifiers |= Pattern.MULTILINE;
                break;
            case 's':
                modifiers |= Pattern.DOTALL;
                break;
            case 'u':
                modifiers |= Pattern.UNICODE_CASE;
                break;
            case 'U':
                modifiers |= Pattern.UNICODE_CHARACTER_CLASS;
                break;
            default:
                break;
        }

        this.add( "" );
        return this;
    }

    public VerbalExpressions removeModifier(char modifier) {
        switch (modifier) {
            case 'd':
                modifiers ^= Pattern.UNIX_LINES;
                break;
            case 'i':
                modifiers ^= Pattern.CASE_INSENSITIVE;
                break;
            case 'x':
                modifiers ^= Pattern.COMMENTS;
                break;
            case 'm':
                modifiers ^= Pattern.MULTILINE;
                break;
            case 's':
                modifiers ^= Pattern.DOTALL;
                break;
            case 'u':
                modifiers ^= Pattern.UNICODE_CASE;
                break;
            case 'U':
                modifiers ^= Pattern.UNICODE_CHARACTER_CLASS;
                break;
            default:
                break;
        }

        this.add( "" );
        return this;
    }

    public VerbalExpressions withAnyCase(boolean enable) {
        if (enable) this.addModifier( 'i' );
        else this.removeModifier( 'i' );
        this.add( "" );
        return this;
    }

    public VerbalExpressions withAnyCase() {
        return withAnyCase(true);
    }

    public VerbalExpressions searchOneLine(boolean enable) {
        if (enable) this.removeModifier( 'm' );
        else this.addModifier( 'm' );
        this.add( "" );
        return this;
    }

    public VerbalExpressions multiple(String value) {
        value = this.sanitize(value);
        switch (value.charAt(0)) {
            case '*':
            case '+':
                break;
            default:
                value += '+';
        }
        this.add(value);
        return this;
    }

    public VerbalExpressions or(String value) {
        if (this.prefixes.indexOf("(") == -1) this.prefixes += "(";
        if (this.suffixes.indexOf(")") == -1) this.suffixes = ")" + this.suffixes;

        this.add( ")|(" );
        if (value != null) this.then(value);
        return this;
    }

    public boolean test(String toTest) {
        this.add( "" );
        return Pattern.matches(this.pattern, toTest);
    }

    public String toString() {
        this.add( "" );
        return this.pattern.toString();
    }
}

File: VerbalExpressions.cs
/*!
 * VerbalExpressions .NET Library v0.1
 * https://github.com/psoholt/VerbalExpressions.net
 *
 *
 * Date: 2013-07-25
 * 
 */

using System.Text.RegularExpressions;

namespace VerbalExpression.Net
{
	public class VerbalExpressions : Regex
	{
		private string _prefixes = "";
		private string _source = "";
		private string _suffixes = "";
		private string _pattern = "";

		private RegexOptions _modifiers = RegexOptions.Multiline;
		private Regex p;

		private string Sanitize(string value)
		{
			if (value != null) 
				return value;
			return Escape(value);
		}

		public VerbalExpressions Add(string value)
		{
			_source = _source != null ? _source + value : value;
			if (_source != null)
			{
				p = new Regex(_prefixes + _source + _suffixes, RegexOptions.Multiline);
				_pattern = p.ToString();
			}
			return this;
		}

		public VerbalExpressions StartOfLine(bool enable)
		{
			_prefixes = enable ? "^" : "";
			return this;
		}

		public VerbalExpressions StartOfLine()
		{
			return StartOfLine(true);
		}

		public VerbalExpressions EndOfLine(bool enable)
		{
			_suffixes = enable ? "$" : "";
			return this;
		}

		public VerbalExpressions EndOfLine()
		{
			return EndOfLine(true);
		}

		public VerbalExpressions Then(string value)
		{
			value = Sanitize(value);
			Add("(" + value + ")");
			return this;
		}

		public VerbalExpressions Find(string value)
		{
			Then(value);
			return this;
		}

		public VerbalExpressions Maybe(string value)
		{
			value = Sanitize(value);
			Add("(" + value + ")?");
			return this;
		}

		public VerbalExpressions Anything()
		{
			Add("(.*)");
			return this;
		}

		public VerbalExpressions AnythingBut(string value)
		{
			value = Sanitize(value);
			Add("([^" + value + "]*)");
			return this;
		}

		public VerbalExpressions Replace(string value)
		{
			_source.Replace(_pattern, value);
			return this;
		}

		public VerbalExpressions LineBreak()
		{
			Add("(\\n|(\\r\\n))");
			return this;
		}

		public VerbalExpressions Br()
		{
			LineBreak();
			return this;
		}

		public VerbalExpressions Tab()
		{
			Add("\\t");
			return this;
		}

		public VerbalExpressions Word()
		{
			Add("\\w+");
			return this;
		}

		public VerbalExpressions AnyOf(string value)
		{
			value = Sanitize(value);
			Add("[" + value + "]");
			return this;
		}

		public VerbalExpressions Any(string value)
		{
			AnyOf(value);
			return this;
		}

		public VerbalExpressions Range(object[] args)
		{
			string value = "[";
			for (int _from = 0; _from < args.Length; _from += 2)
			{
				int _to = _from + 1;
				if (args.Length <= _to) break;
				string from = Sanitize((string)args[_from]);
				string to = Sanitize((string)args[_to]);

				value += from + "-" + to;
			}

			value += "]";

			Add(value);
			return this;
		}

		public VerbalExpressions AddModifier(char modifier)
		{
			switch (modifier)
			{
				//case 'd':
				//	_modifiers |= RegexOptions.UNIX_LINES;
				//	break;
				case 'i':
					_modifiers |= RegexOptions.IgnoreCase;
					break;
				//case 'x':
				//	_modifiers |= Pattern.COMMENTS;
				//	break;
				case 'm':
					_modifiers |= RegexOptions.Multiline;
					break;
				//case 's':
				//	_modifiers |= RegexOptions.DOTALL;
				//	break;
				//case 'u':
				//	_modifiers |= Pattern.UNICODE_CASE;
				//	break;
				//case 'U':
				//	_modifiers |= Pattern.UNICODE_CHARACTER_CLASS;
				//	break;
			}

			return this;
		}

		public VerbalExpressions RemoveModifier(char modifier)
		{
			switch (modifier)
			{
				//case 'd':
				//	_modifiers ^= Pattern.UNIX_LINES;
				//	break;
				case 'i':
					_modifiers ^= RegexOptions.IgnoreCase;
					break;
				//case 'x':
				//	_modifiers ^= Pattern.COMMENTS;
				//	break;
				case 'm':
					_modifiers ^= RegexOptions.Multiline;
					break;
				//case 's':
				//	_modifiers ^= Pattern.DOTALL;
				//	break;
				//case 'u':
				//	_modifiers ^= Pattern.UNICODE_CASE;
				//	break;
				//case 'U':
				//	_modifiers ^= Pattern.UNICODE_CHARACTER_CLASS;
				//	break;
				//default:
				//	break;
			}

			return this;
		}

		public VerbalExpressions WithAnyCase(bool enable)
		{
			if (enable) AddModifier('i');
			else RemoveModifier('i');
			return this;
		}

		public VerbalExpressions WithAnyCase()
		{
			return WithAnyCase(true);
		}

		public VerbalExpressions SearchOneLine(bool enable)
		{
			if (enable) RemoveModifier('m');
			else AddModifier('m');
			return this;
		}

		public VerbalExpressions Multiple(string value)
		{
			value = Sanitize(value);
			switch (value[0])
			{
				case '*':
				case '+':
					break;
				default:
					value += '+';
				break;
			}

			Add(value);
			return this;
		}

		public VerbalExpressions Or(string value)
		{
			if (_prefixes.IndexOf("(") == -1) _prefixes += "(";
			if (_suffixes.IndexOf(")") == -1) _suffixes = ")" + _suffixes;

			Add(")|(");
			if (value != null) Then(value);
			return this;
		}

		public bool Test(string toTest)
		{
			return IsMatch(toTest, _pattern);
		}

		public override string ToString()
		{
			return _pattern;
		}
	}
}
